File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/mylist/MylistFragment.java
package com.twtstudio.tjwhm.lostfound.mylist;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.twtstudio.tjwhm.lostfound.R;
import com.twtstudio.tjwhm.lostfound.R2;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by tjwhm on 2017/7/4.
 **/

public class MylistFragment extends Fragment implements MylistContract.MylistView {

    @BindView(R2.id.mylist_recyclerView)
    RecyclerView mylist_recyclerView;
    @BindView(R2.id.mylist_progress)
    ProgressBar mylist_progress;
    @BindView(R2.id.mylist_nodata)
    LinearLayout mylist_nodata;
    boolean isLoading = false;
    boolean needClear = false;

    private MylistTableAdapter tableAdapter;
    private LinearLayoutManager layoutManager;
    private MylistBean mylistBean;
    private String lostOrFound;
    MylistContract.MylistPresenter mylistPresenter = new MylistPresenterImpl(this);
    int page = 1;

    public static MylistFragment newInstance(String type) {

        Bundle args = new Bundle();
        args.putString("index", type);
        MylistFragment fragment = new MylistFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.lf_fragment_mylist, container, false);
        ButterKnife.bind(this, view);
        Bundle bundle = getArguments();
        lostOrFound = bundle.getString("index");
        initValues();

        mylist_recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int totalCount = layoutManager.getItemCount();
                int lastVisibleItem = layoutManager.findLastCompletelyVisibleItemPosition();
                if (!isLoading && totalCount < (lastVisibleItem + 2)) {
                    ++page;
                    isLoading = true;
                    mylistPresenter.loadMylistData(lostOrFound, page);
                }
            }
        });
        return view;
    }

    @Override
    public void setMylistData(MylistBean mylistBean) {
        if (needClear) {
            this.mylistBean.data.clear();
        }
        this.mylistBean.error_code = mylistBean.error_code;
        this.mylistBean.message = mylistBean.message;
        this.mylistBean.data.addAll(mylistBean.data);
        tableAdapter.notifyDataSetChanged();
        mylist_progress.setVisibility(View.GONE);
        if (this.mylistBean.data.size() == 0 && page == 1) {
            mylist_nodata.setVisibility(View.VISIBLE);
        } else {
            mylist_nodata.setVisibility(View.GONE);
        }
        isLoading = false;
        needClear = false;
    }

    @Override
    public void turnStatus(int id) {
        mylistPresenter.turnStatus(id);
    }

    @Override
    public void turnStatusSuccessCallBack() {
        needClear = true;
        mylistPresenter.loadMylistData(lostOrFound, 1);
    }

    private void initValues() {
        mylist_nodata.setVisibility(View.GONE);
        mylist_progress.setVisibility(View.VISIBLE);
        mylistBean = new MylistBean();
        mylistBean.data = new ArrayList<>();
        layoutManager = new LinearLayoutManager(getActivity());
        mylist_recyclerView.setLayoutManager(layoutManager);
        tableAdapter = new MylistTableAdapter(mylistBean, getActivity(), lostOrFound, this);
        mylist_recyclerView.setAdapter(tableAdapter);
    }

    @Override
    public void onResume() {
        super.onResume();
        page = 1;
        mylistBean.data = new ArrayList<>();
        tableAdapter.notifyDataSetChanged();
        mylistPresenter.loadMylistData(lostOrFound, page);

    }
}

File: WePeiYang/lostfound2/src/main/java/com/yookiely/lostfond2/mylist/MyListFragment.kt
package com.yookiely.lostfond2.mylist

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.Toast
import com.example.lostfond2.R
import com.yookiely.lostfond2.service.MyListDataOrSearchBean
import com.yookiely.lostfond2.service.Utils
import es.dmoral.toasty.Toasty
import org.jetbrains.anko.support.v4.toast


class MyListFragment : Fragment(), MyListService.MyListView {

    private lateinit var myListRecyclerView: RecyclerView
    private lateinit var myListProgressBar: ProgressBar
    private lateinit var myListNoData: LinearLayout
    var isLoading = false
    private var needClear = false
    private lateinit var tableAdapter: MyListTableAdapter
    lateinit var myListLayoutManager: LinearLayoutManager
    private var myListBean: MutableList<MyListDataOrSearchBean> = mutableListOf()// 可能会有bug
    lateinit var lostOrFound: String
    private val myListPresenter: MyListService.MyListPresenter = MyListPresenterImpl(this)
    var page = 1

    companion object {
        fun newInstance(type: String): MyListFragment {
            val args = Bundle()
            args.putString(Utils.INDEX_KEY, type)
            val fragment = MyListFragment()
            fragment.arguments = args
            return fragment
        }
    }

    override fun setMyListData(myListBean: List<MyListDataOrSearchBean>) {
        if (page==1 || needClear ) this.myListBean.clear()
        this.myListBean.addAll(myListBean)
        tableAdapter.notifyDataSetChanged()
        myListProgressBar.visibility = View.GONE
        myListNoData.visibility = if (tableAdapter.myListBean.size == 0 && page == 1) View.VISIBLE else View.GONE
        isLoading = false
        needClear = false
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view: View = inflater.inflate(R.layout.lf2_fragment_mylist, container, false)
        myListRecyclerView = view.findViewById(R.id.rv_mylist_recyclerView)
        myListProgressBar = view.findViewById(R.id.pb_mylist_progress)
        myListNoData = view.findViewById(R.id.ll_mylist_nodata)
        val bundle = arguments
        if (bundle == null ){
            Toasty.error(this.context!!,"没有拿到数据",Toast.LENGTH_SHORT,true).show()
        }else{
            lostOrFound = bundle.getString(Utils.INDEX_KEY)
        }
        initValues()

        myListRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val totalCount = myListLayoutManager.itemCount
                val lastVisibleItem = myListLayoutManager.findLastCompletelyVisibleItemPosition()
                if (!isLoading && totalCount < (lastVisibleItem + 2) && totalCount > 4) {
                    ++page
                    isLoading = true
                    MyListPresenterImpl(this@MyListFragment).loadMyListData(lostOrFound, page)
                }
            }

        })

        return view
    }

    override fun onResume() {
        super.onResume()
        if (Utils.needRefresh){
            page = 1
            myListPresenter.loadMyListData(lostOrFound,page)
            if (lostOrFound == "lost") Utils.needRefresh = false
        }
    }

    override fun turnStatus(id: Int) {
        myListPresenter.turnStatus(id)
    }

    override fun turnStatusSuccessCallBack() {
        needClear = true
        myListPresenter.loadMyListData(lostOrFound, 1)
    }

    private fun initValues() {
        myListNoData.visibility = View.GONE
        myListProgressBar.visibility = View.VISIBLE

        myListLayoutManager = object : LinearLayoutManager(activity) {
            override fun generateDefaultLayoutParams(): RecyclerView.LayoutParams = RecyclerView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        }
        myListLayoutManager.orientation = LinearLayoutManager.VERTICAL
        tableAdapter = MyListTableAdapter(myListBean, activity, lostOrFound, this)
        myListRecyclerView.apply {
            layoutManager = myListLayoutManager
            adapter = tableAdapter
        }
        myListPresenter.loadMyListData(lostOrFound, page)
    }
}
